### ISSUE TITLE
Shape mismatch while unpooling

### ISSUE THREAD

**Issue by fabianbormann** [fabianbormann/Tensorflow_Segmentation#6]
The Error appears in line 147 respectively in line 234. Maybe there is some flooring or ceiling during the input size and or conv layers. 

Steps to reproduce:
```zsh
cd /your/path/to/Tensorflow-DeconvNet-Segmentation/
python3 main.py
```

> Error: Caused by op 'concat_1', defined at:
  File "main.py", line 4, in <module>
    deconvNet = DeconvNet(use_cpu=True)
  File "/home/fabian/workspace/Tensorflow-DeconvNet-Segmentation/DeconvNet.py", line 15, in __init__
    self.build(use_cpu=use_cpu)
  File "/home/fabian/workspace/Tensorflow-DeconvNet-Segmentation/DeconvNet.py", line 147, in build
    unpool_4 = self.unpool_layer2x2(deconv_5_1, self.unravel_argmax(pool_4_argmax, tf.to_int64(tf.shape(conv_4_3))))
  File "/home/fabian/workspace/Tensorflow-DeconvNet-Segmentation/DeconvNet.py", line 234, in unpool_layer2x2
    t = tf.concat(3, [t2, t1])
  File "/usr/local/lib/python3.5/dist-packages/tensorflow/python/ops/array_ops.py", line 872, in concat
    name=name)
  File "/usr/local/lib/python3.5/dist-packages/tensorflow/python/ops/gen_array_ops.py", line 436, in _concat
    values=values, name=name)
  File "/usr/local/lib/python3.5/dist-packages/tensorflow/python/framework/op_def_library.py", line 749, in apply_op
    op_def=op_def)
  File "/usr/local/lib/python3.5/dist-packages/tensorflow/python/framework/ops.py", line 2380, in create_op
    original_op=self._default_original_op, op_def=op_def)
  File "/usr/local/lib/python3.5/dist-packages/tensorflow/python/framework/ops.py", line 1298, in __init__
    self._traceback = _extract_stack()

**Comment by AngusG** [comment]
Why do you use three channels here? This doesn't seem logical to me for a segmentation mask, unless I missed something in the original paper.
`self.y = tf.placeholder(tf.int64, shape=(1, None, None, 3))`

Also, you can remove the opencv dependency and have tensorflow automatically read and crop the images to an even 224x224 with the following:

```
        train_images_file_list=[]
        train_masks_file_list=[]
        trainset = open('data/stage_1_train_imgset/train.txt').readlines()
        for line in trainset:
            train_images_file_list.append(line.split(' ')[0])
            train_masks_file_list.append(line.split(' ')[1][:-1])

        train_image_filename_queue = tf.train.string_input_producer(train_images_file_list)
        train_mask_filename_queue = tf.train.string_input_producer(train_masks_file_list)

        train_images = tf.image.decode_jpeg(train_img_content, channels=3)
        train_masks = tf.image.decode_jpeg(train_mask_content, channels=1)

        train_images = tf.cast(train_images, tf.float32)
        train_resized_images=tf.image.resize_image_with_crop_or_pad(train_images, 224, 224)
        train_resized_masks=tf.image.resize_image_with_crop_or_pad(train_masks, 224, 224)

        # create mini-batch
        train_image_batch = tf.train.batch([train_resized_images], batch_size=bs)
        train_masks_batch = tf.train.batch([train_resized_masks], batch_size=bs)

```

I was about to submit a PR after getting this to run on my own dataset, but began running into issues when attempting to generalize to original VOC2012 data. If you try to pass the `train_image_batch` directly to `feed_dict` in `self.train_step.run()` you get the ValueError:

> setting an array element with a sequence

since `train_image_batch` is a tensor object. After removing the placeholders and setting `self.x` and `self.y` directly as `train_image_batch` and `train_masks_batch`, I am getting shape errors in the `unpool_layer`s.

**Comment by fabianbormann** [comment]
Hi @AngusG, 

yes you are completely right! I copied the train method from my master thesis (where I need rgb images as ground truth) and started to refactor the method for this segmentation problem. But I endet up in the Shape mismatch first -> that's why the image reading (and OpenCV dependencies) seems to be a little bit strange ... So please submit your PR :+1: it's very welcome!! 

I will try to figure out this bad shape bug asap!
If you want I could give you directly write access to this repo?

Cheers,
Fabian

**Comment by AngusG** [comment]
Thanks for the quick follow up, i'd be curious to have a look at your thesis too! Feel free to add me to the repo, i'll see if I can push a working example with a subset of my data to a separate branch.

**Comment by fabianbormann** [comment]
Hi @AngusG, I solved the shape problem (currently only in the [UnpoolLayerTest](https://github.com/fabianbormann/Tensorflow-DeconvNet-Segmentation/blob/master/tests/UnpoolLayerTest.ipynb)) should we merge the branches first?

The solution was only to use a ceiling instead of flooring in the computation so ```width // 2``` becomes ```(width + 1) // 2``` and ```unravel_argmax``` should be inside of the unpool layer method, because we need to use the ```output_shape``` also inside of it.

```python
def unpool_layer2x2(x, raveled_argmax, out_shape):
    argmax = unravel_argmax(raveled_argmax, tf.to_int64(out_shape))

    # use
    output = tf.zeros([out_shape[1], out_shape[2], out_shape[3]])
    # don't compute the shape like this
    # x_shape = tf.shape(x)
    # output = tf.zeros([x_shape[1] * 2, x_shape[2] * 2, x_shape[3]])
```

**Comment by AngusG** [comment]
Nice find! I confirmed its working. I'm thinking we should also offer a TFRecord with mini-batch implementation so training converges faster. Should I just make a new file in master? This is what I had started in `prefetch`. At the moment the two branches won't merge well, so this will need some work.

**Comment by fabianbormann** [comment]
Yeah nice idea -> Do it :+1: ! Meanwhile I want to figure out, if we could use/convert some pretrained coffe weights .. I think it would be great if we provide a trained model. I read that [there is a way](https://github.com/ethereon/caffe-tensorflow) to convert caffe to tensorflow weights. [HyeonwooNoh](https://github.com/HyeonwooNoh/DeconvNet) use the FCN-8s model and weight file in their implementation. Maybe we could use it too? Do you have any experience with TF <-> Caffe weight conversions?

**Comment by AngusG** [comment]
Agreed that's a great idea. I can look at it more in the next few days, in the meantime check out https://github.com/MarvinTeichmann/tensorflow-fcn
